File: 2021/day_4.js
// https://adventofcode.com/2021/day/4

const assert = require('assert');

const [DRAWS, BOARDS] = (() => {
  let LINES = require('fs')
    .readFileSync(__dirname + '/day_4.txt', 'utf8')
    .trimEnd()
    .split('\n\n');

  let DRAWS = LINES.shift()
    .split(',')
    .map(x => parseInt(x));

  const BOARDS = LINES.map(x =>
    x.split('\n').map(x =>
      x
        .split(' ')
        .filter(x => x.length > 0)
        .map(x => [parseInt(x), false]),
    ),
  );

  return [DRAWS, BOARDS];
})();

function did_win(board) {
  // Rows.
  for (let i = 0; i < board.length; i++) {
    if (board[i].every(cell => cell[1])) {
      return true;
    }
  }

  // Columns.
  for (let i = 0; i < board[0].length; i++) {
    if (board.every(row => row[i][1])) {
      return true;
    }
  }

  return false;
}

function sum_unmarked(board) {
  let sum = 0;
  for (const row of board) {
    for (const cell of row) {
      if (cell[1] === false) {
        sum += cell[0];
      }
    }
  }
  return sum;
}

(function part_1() {
  for (const draw of DRAWS) {
    for (const board of BOARDS) {
      for (const row of board) {
        for (const cell of row) {
          if (cell[0] == draw) {
            cell[1] = true;
          }
        }
      }
      if (did_win(board)) {
        const answer = sum_unmarked(board) * draw;
        console.log('part 1: %s', answer);
        assert.equal(answer, 54275);
        return;
      }
    }
  }
})();

(function part_2() {
  const wins = [];

  for (const draw of DRAWS) {
    for (let i = BOARDS.length - 1; i >= 0; i--) {
      let board = BOARDS[i];

      for (const row of board) {
        for (const cell of row) {
          if (cell[0] == draw) {
            cell[1] = true;
          }
        }
      }

      if (did_win(board)) {
        wins.push([board, draw]);
        BOARDS.splice(i, 1);
      }
    }
  }

  let last_win = wins[wins.length - 1];

  const answer = sum_unmarked(last_win[0]) * last_win[1];
  console.log('part 2: %s', answer);
  assert.equal(answer, 13158);
})();

File: 2021/day_4.rs
// https://adventofcode.com/2021/day/4

type Draws = Vec<usize>;
type Board = Vec<Vec<(usize, bool)>>;

fn data() -> (Draws, Vec<Board>) {
    let mut lines_it = include_str!("day_4.txt")
        .trim_end_matches('\n')
        .split("\n\n");

    let draws = lines_it
        .next()
        .unwrap()
        .split(',')
        .map(|x| x.parse::<usize>().unwrap())
        .collect::<Draws>();

    let boards = lines_it
        .map(|x| {
            x.split('\n')
                .map(|x| {
                    x.split(' ')
                        .filter(|x| !x.is_empty())
                        .map(|x| (x.parse::<usize>().unwrap(), false))
                        .collect::<Vec<_>>()
                })
                .collect::<Vec<_>>()
        })
        .collect::<Vec<Board>>();

    (draws, boards)
}

fn did_win(board: &Board) -> bool {
    for row in board {
        if row.iter().all(|x| x.1) {
            return true;
        }
    }

    for col_num in 0..board[0].len() {
        if board.iter().all(|row| row[col_num].1) {
            return true;
        }
    }

    return false;
}

fn sum_unmarked(board: &Board) -> usize {
    board
        .iter()
        .flat_map(|row| row)
        .map(|cell| if cell.1 == false { cell.0 } else { 0 })
        .sum()
}

fn main() {
    // Part 1
    {
        let (draws, mut boards) = data();

        'outer: for draw in &draws {
            for board in boards.as_mut_slice() {
                for row in board.as_mut_slice() {
                    for cell in row {
                        if cell.0 == *draw {
                            cell.1 = true;
                        }
                    }
                }
                if did_win(&board) {
                    let answer = *draw * sum_unmarked(board);
                    println!("part 1: {}", answer);
                    assert_eq!(answer, 54275);
                    break 'outer;
                }
            }
        }
    }

    // Part 2
    {
        let (draws, mut boards) = data();
        let mut wins: Vec<(Board, usize)> = Vec::new();

        for draw in &draws {
            for i in (0..boards.len()).rev() {
                for row in boards[i].as_mut_slice() {
                    for cell in row {
                        if cell.0 == *draw {
                            cell.1 = true;
                        }
                    }
                }

                if did_win(&boards[i]) {
                    let board = boards.remove(i);
                    wins.push((board, *draw));
                }
            }
        }

        let (board, draw) = wins.last().unwrap();
        let answer = draw * sum_unmarked(board);
        println!("part 2: {}", answer);
        assert_eq!(answer, 13158);
    }
}
